File: src/main/java/at/xirado/bean/command/SlashCommand.java
package at.xirado.bean.command;

import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.Command;
import net.dv8tion.jda.api.interactions.commands.CommandPermission;
import net.dv8tion.jda.api.interactions.commands.build.CommandData;
import net.dv8tion.jda.api.interactions.commands.build.SlashCommandData;

import javax.annotation.Nonnull;
import java.util.Arrays;
import java.util.EnumSet;
import java.util.HashSet;
import java.util.Set;

public abstract class SlashCommand implements GenericCommand
{
    private SlashCommandData commandData = null;

    private final EnumSet<Permission> requiredUserPermissions = EnumSet.noneOf(Permission.class);
    private final EnumSet<Permission> requiredBotPermissions = EnumSet.noneOf(Permission.class);
    private final Set<Long> enabledGuilds = new HashSet<>();
    private final EnumSet<CommandFlag> commandFlags = EnumSet.noneOf(CommandFlag.class);

    public void setCommandData(SlashCommandData commandData)
    {
        commandData.setGuildOnly(true);
        this.commandData = commandData;
    }

    public void addRequiredBotPermissions(Permission... permissions)
    {
        requiredBotPermissions.addAll(Arrays.asList(permissions));
    }

    public void addRequiredUserPermissions(Permission... permissions)
    {
        commandData.setDefaultPermissions(CommandPermission.enabledFor(permissions));
        requiredUserPermissions.addAll(Arrays.asList(permissions));
    }

    public void addCommandFlags(CommandFlag... flags)
    {
        commandFlags.addAll(Arrays.asList(flags));
    }

    public void addEnabledGuilds(Long... ids)
    {
        enabledGuilds.addAll(Arrays.asList(ids));
    }

    @Override
    public CommandData getCommandData()
    {
        return commandData;
    }

    @Override
    public EnumSet<Permission> getRequiredUserPermissions()
    {
        return requiredUserPermissions;
    }

    @Override
    public EnumSet<Permission> getRequiredBotPermissions()
    {
        return requiredBotPermissions;
    }

    @Override
    public Command.Type getType()
    {
        return Command.Type.SLASH;
    }

    @Override
    public Set<Long> getEnabledGuilds()
    {
        return enabledGuilds;
    }

    @Override
    public EnumSet<CommandFlag> getCommandFlags()
    {
        return commandFlags;
    }

    /**
     * Executes requested slash command
     *
     * @param event  The SlashCommandInteractionEvent
     * @param ctx    Helpful methods in context of the event
     */
    public abstract void executeCommand(@Nonnull SlashCommandInteractionEvent event, @Nonnull SlashCommandContext ctx);

    public void handleAutocomplete(@Nonnull CommandAutoCompleteInteractionEvent event) throws Exception {};

}

File: src/main/kotlin/at/xirado/bean/interaction/SlashCommand.kt
package at.xirado.bean.interaction

import dev.minn.jda.ktx.interactions.Subcommand
import dev.minn.jda.ktx.interactions.optionType
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.Command
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.Commands
import net.dv8tion.jda.api.interactions.commands.build.OptionData
import net.dv8tion.jda.api.interactions.commands.build.SubcommandData
import java.util.*

abstract class SlashCommand(name: String, description: String) : GenericCommand {
    override val commandData = Commands.slash(name.lowercase(Locale.getDefault()), description)
    override val requiredUserPermissions: EnumSet<Permission> = EnumSet.noneOf(Permission::class.java)
    override val requiredBotPermissions: EnumSet<Permission> = EnumSet.noneOf(Permission::class.java)
    override val enabledGuilds = HashSet<Long>()
    override val commandFlags: EnumSet<CommandFlag> = EnumSet.noneOf(CommandFlag::class.java)
    override var disabled: Boolean = false

    override val type: Command.Type
        get() = Command.Type.SLASH

    fun devCommand(requiresAdmin: Boolean = true) {
        enabledGuilds.addAll(application.config.devGuilds)
        commandFlags.add(CommandFlag.DEVELOPER_ONLY)
        if (requiresAdmin)
            requiredUserPermissions.add(Permission.ADMINISTRATOR)
    }

    inline fun <reified T> option(name: String, description: String, required: Boolean = false, autocomplete: Boolean = false, builder: OptionData.() -> Unit = {}) {
        val type = optionType<T>()
        if (type == OptionType.UNKNOWN)
            throw IllegalArgumentException("Cannot resolve type " + T::class.java.simpleName + " to OptionType!")

        val nameLocalizations = application.localizationManager.getDiscordLocalizations("commands.slash.${commandData.name}.options.$name.name")
        val descriptionLocalizations = application.localizationManager.getDiscordLocalizations("commands.slash.${commandData.name}.options.$name.description")

        commandData.addOptions(
            OptionData(type, name, description)
                .setRequired(required)
                .setAutoComplete(autocomplete)
                .setNameLocalizations(nameLocalizations)
                .setDescriptionLocalizations(descriptionLocalizations)
                .apply(builder)
        )
    }

    inline fun subCommand(name: String, description: String, builder: SubcommandData.() -> Unit = {}) = commandData.addSubcommands(
        Subcommand(name, description, builder)
    )

    open suspend fun baseCommand(event: SlashCommandInteractionEvent) {}

}